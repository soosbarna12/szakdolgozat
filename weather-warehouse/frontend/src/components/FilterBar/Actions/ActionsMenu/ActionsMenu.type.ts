export interface ActionsMenuProps {
    anchorElUser: null | HTMLElement;
    handleCloseMenu: (event: React.MouseEvent<HTMLElement>) => void;
    onSaveLocation?: () => void;
    onResetLocation?: () => void;
}
