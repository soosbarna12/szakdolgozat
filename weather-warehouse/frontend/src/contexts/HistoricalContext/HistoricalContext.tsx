import React, { createContext, useEffect, useState } from "react";
import { LOCAL_STORAGE_TEMPERATURE_SCALE } from "../../consts/temperatureScale.const";
import { TemperatureScale } from "../../types/temperatureScale.type";
import { HistoricalContextType, HistoricalLocationData, HistoricalPageData, HistoricalProviderProps } from "./HistoricalContext.type";


export const HistoricalContext = createContext<HistoricalContextType>({
  location: { name: "", lat: 0, lon: 0 },
  setLocation: () => { },
  historicalPageData: [] as any,
  setHistoricalPageData: () => { },
  temperatureScale: TemperatureScale.Celsius,
  setTemperatureScale: () => { },
});

export const HistoricalLocationProvider = ({ children }: HistoricalProviderProps) => {

  const [location, setLocation] = useState<HistoricalLocationData>(() => {
    const storedLocation = localStorage.getItem("historicalLocation");
    const defaultLocation = { name: "", lat: 0, lon: 0 };
    try {
      return storedLocation ? JSON.parse(storedLocation) : defaultLocation;
    } catch (error) {
      console.error("Failed to parse location from localStorage:", error);
      return defaultLocation;
    }
  });

  const [historicalPageData, setHistoricalPageData] = useState<HistoricalPageData[]>(() => {
    const storedData = localStorage.getItem("historicalPageData");
    try {
      return storedData ? JSON.parse(storedData) : [];
    } catch (error) {
      console.error("Failed to parse historical data from localStorage:", error);
      return [];
    }
  });

  const [temperatureScale, setTemperatureScale] = useState<TemperatureScale | string>(() => {
    const storedTemperatureScale = localStorage.getItem(LOCAL_STORAGE_TEMPERATURE_SCALE);
    try {
      return storedTemperatureScale ? JSON.parse(storedTemperatureScale) : TemperatureScale.Celsius;
    } catch (error) {
      console.error("Failed to parse temperature scale from localStorage:", error);
      return TemperatureScale.Celsius;
    }
  });


  useEffect(() => {
    if (historicalPageData) {
      localStorage.setItem("historicalPageData", JSON.stringify(historicalPageData));
    }
  }, [historicalPageData]);

  useEffect(() => {
    localStorage.setItem("historicalLocation", JSON.stringify(location));
  }, [location]);

  useEffect(() => {
    localStorage.setItem(LOCAL_STORAGE_TEMPERATURE_SCALE, JSON.stringify(temperatureScale));
  }, [temperatureScale]);


  return (
    <HistoricalContext.Provider value={{ location, setLocation, historicalPageData, setHistoricalPageData, temperatureScale, setTemperatureScale }}>
      {children}
    </HistoricalContext.Provider>
  );
};
