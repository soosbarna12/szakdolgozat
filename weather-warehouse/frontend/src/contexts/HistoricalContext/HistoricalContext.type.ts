import { Dispatch, ReactNode, SetStateAction } from "react";
import { HistoricalDataTable } from "../../types/historicalDataTable.type";
import { TemperatureScale } from "../../types/temperatureScale.type";

export interface HistoricalLocationData {
  name: string;
  country?: string;
  lat: number;
  lon: number;
}

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface HistoricalPageData extends HistoricalDataTable {}

export interface HistoricalContextType {
  location: HistoricalLocationData;
  historicalPageData: any [];
  temperatureScale: TemperatureScale | string;
  setLocation: Dispatch<SetStateAction<HistoricalLocationData>>;
  setHistoricalPageData: Dispatch<SetStateAction<HistoricalPageData[]>>;
  setTemperatureScale: Dispatch<SetStateAction<TemperatureScale | string>>;
}

export interface HistoricalProviderProps {
  children: ReactNode;
}
