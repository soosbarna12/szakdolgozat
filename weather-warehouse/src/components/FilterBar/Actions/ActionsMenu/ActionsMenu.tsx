import { MenuItem, Typography } from "@mui/material";
import React from "react";
import { StyledMenu } from "../../../../common.style";
import { ActionsMenuProps } from "./ActionsMenu.type";

const ActionsMenuItems = ["Edit", "Duplicate", "Archive", "More"];

export function ActionsMenu(props: Readonly<ActionsMenuProps>) {
  const { anchorElUser, handleCloseMenu } = props;

  return (
    <StyledMenu
      sx={{ mt: "45px", padding: "0", borderRadius: "20px" }}
      id="menu-appbar"
      anchorEl={anchorElUser}
      anchorOrigin={{ vertical: "top", horizontal: "right", }}
      keepMounted
      transformOrigin={{ vertical: "top", horizontal: "right", }}
      open={Boolean(anchorElUser)}
      onClose={handleCloseMenu}
    >
      {ActionsMenuItems.map((ActionsMenuItems) => (
        <MenuItem key={ActionsMenuItems} onClick={handleCloseMenu}>
          <Typography sx={{ textAlign: "center" }}>{ActionsMenuItems}</Typography>
        </MenuItem>
      ))}
    </StyledMenu >
  );
}