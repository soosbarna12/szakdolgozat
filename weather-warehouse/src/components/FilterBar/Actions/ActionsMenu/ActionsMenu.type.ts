export interface ActionsMenuProps {
	anchorElUser: null | HTMLElement;
	handleCloseMenu: (event: React.MouseEvent<HTMLElement>) => void;
}
