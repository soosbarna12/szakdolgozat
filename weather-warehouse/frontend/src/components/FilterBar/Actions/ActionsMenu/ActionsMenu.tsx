import { Typography } from "@mui/material";
import React, { useContext } from "react";
import { HistoricalContext } from "../../../../contexts/HistoricalContext/HistoricalContext";
import { StyledMenu, StyledMenuItem } from "../../../../stlyes/common.style";
import { ActionsMenuProps } from "./ActionsMenu.type";

export function ActionsMenu(props: Readonly<ActionsMenuProps>) {
  const { anchorElUser, handleCloseMenu, onSaveLocation, onExportLocation, onResetLocation } = props;
  const { historicalPageData } = useContext(HistoricalContext);

  function handleSaveClick() {
    onSaveLocation?.();
    handleCloseMenu(null as any);
  };

  function handleExportClick() {
    onExportLocation?.();
    handleCloseMenu(null as any);
  };

  function handleResetClick() {
    onResetLocation?.();
    handleCloseMenu(null as any);
  };


  return (
    <StyledMenu
      PaperProps={{ sx: { width: "120px" } }}
      sx={{ mt: "45px", padding: "0", borderRadius: "20px" }}
      id="menu-appbar"
      anchorEl={anchorElUser}
      anchorOrigin={{ vertical: "top", horizontal: "right" }}
      keepMounted
      transformOrigin={{ vertical: "top", horizontal: "right" }}
      open={Boolean(anchorElUser)}
      onClose={handleCloseMenu}
    >
      <StyledMenuItem data-testid="saveMenuItem" key={"save"} onClick={handleSaveClick} disabled={historicalPageData?.length === 0}>
        <Typography sx={{ textAlign: "center" }}>Save</Typography>
      </StyledMenuItem>
      <StyledMenuItem data-testid="exportMenuItem" key={"export"} onClick={handleExportClick} disabled={historicalPageData?.length === 0}>
        <Typography sx={{ textAlign: "center" }}>Export</Typography>
      </StyledMenuItem>
      <StyledMenuItem data-testid="resetMenuItem" key={"reset"} onClick={handleResetClick} disabled={historicalPageData?.length === 0}>
        <Typography sx={{ textAlign: "center" }}>Reset</Typography>
      </StyledMenuItem>
    </StyledMenu>
  );
}
