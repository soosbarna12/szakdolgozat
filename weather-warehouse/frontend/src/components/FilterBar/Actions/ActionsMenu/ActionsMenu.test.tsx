import '@testing-library/jest-dom';
import { fireEvent, screen } from '@testing-library/react';
import React from 'react';
import { HistoricalContext } from '../../../../contexts/HistoricalContext/HistoricalContext';
import { renderWithQueryClient } from '../../../../utils/test/renderWithQueryClient';
import { ActionsMenu } from './ActionsMenu';

describe('FilterBar/Actions/ActionsMenu', () => {
    const mockHandleCloseMenu = jest.fn();
    const mockOnSaveLocation = jest.fn();
    const mockOnExportLocation = jest.fn();
    const mockOnResetLocation = jest.fn();

    const renderComponent = (historicalPageData: any[] = [], anchorElUser: HTMLElement | null = document.createElement('div')) => {
        return renderWithQueryClient(
            <HistoricalContext.Provider value={{ historicalPageData } as any}>
                <ActionsMenu
                    anchorElUser={anchorElUser}
                    handleCloseMenu={mockHandleCloseMenu}
                    onSaveLocation={mockOnSaveLocation}
                    onExportLocation={mockOnExportLocation}
                    onResetLocation={mockOnResetLocation}
                />
            </HistoricalContext.Provider>
        );
    };

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('matches ActionsMenu component snapshot', () => {
        const component = renderComponent();
        expect(component).toMatchSnapshot();
    });

    it('renders menu items and handles Save button click', () => {
        renderComponent([{ id: 1 }]); // Simulate historicalPageData with one item

        const saveButton = screen.getByTestId('saveMenuItem');
        expect(saveButton).not.toBeDisabled();

        fireEvent.click(saveButton);

        expect(mockOnSaveLocation).toHaveBeenCalled();
        expect(mockHandleCloseMenu).toHaveBeenCalledWith(null);
    });

    it('handles Export button click', () => {
        renderComponent([{ id: 1 }]); // Simulate historicalPageData with one item

        const exportButton = screen.getByTestId('exportMenuItem');
        expect(exportButton).not.toBeDisabled();

        fireEvent.click(exportButton);

        expect(mockOnExportLocation).toHaveBeenCalled();
        expect(mockHandleCloseMenu).toHaveBeenCalledWith(null);
    });

    it('handles Reset button click', () => {
        renderComponent([{ id: 1 }]); // Simulate historicalPageData with one item

        const resetButton = screen.getByTestId('resetMenuItem');
        expect(resetButton).not.toBeDisabled();

        fireEvent.click(resetButton);

        expect(mockOnResetLocation).toHaveBeenCalled();
        expect(mockHandleCloseMenu).toHaveBeenCalledWith(null);
    });

    it('disables menu items when historicalPageData is empty', () => {
        renderComponent([]); // Simulate empty historicalPageData

        const saveButton = screen.getByTestId('saveMenuItem');
        const exportButton = screen.getByTestId('exportMenuItem');
        const resetButton = screen.getByTestId('resetMenuItem');

        expect(saveButton).toHaveAttribute('aria-disabled', 'true');
        expect(exportButton).toHaveAttribute('aria-disabled', 'true');
        expect(resetButton).toHaveAttribute('aria-disabled', 'true');
    });

    it('does not render menu when anchorElUser is null', () => {
        renderComponent([], null); // Simulate no anchor element

        expect(screen.queryByTestId('actionsMenu')).not.toBeInTheDocument();
    });
});