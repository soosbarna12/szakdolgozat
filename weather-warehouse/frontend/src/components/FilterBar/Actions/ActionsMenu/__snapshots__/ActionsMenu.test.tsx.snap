// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBar/Actions/ActionsMenu matches ActionsMenu component snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="MuiPopover-root MuiMenu-root MuiModal-root css-6eyezz-MuiModal-root-MuiPopover-root-MuiMenu-root"
      id="menu-appbar"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-vl6bb4-MuiBackdrop-root-MuiModal-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiPopover-paper MuiMenu-paper css-t3bg0k-MuiPaper-root-MuiPopover-paper-MuiMenu-paper"
        style="--Paper-shadow: 0px 5px 5px -3px rgba(0,0,0,0.2),0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12); opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
        tabindex="-1"
      >
        <ul
          class="MuiList-root MuiList-padding MuiMenu-list css-1toxriw-MuiList-root-MuiMenu-list"
          role="menu"
          tabindex="0"
        >
          <li
            aria-disabled="true"
            class="MuiButtonBase-root MuiMenuItem-root Mui-disabled MuiMenuItem-gutters Mui-disabled MuiMenuItem-root Mui-disabled MuiMenuItem-gutters css-nl63xq-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="saveMenuItem"
            role="menuitem"
            tabindex="-1"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1xrrpwr-MuiTypography-root"
            >
              Save
            </p>
          </li>
          <li
            aria-disabled="true"
            class="MuiButtonBase-root MuiMenuItem-root Mui-disabled MuiMenuItem-gutters Mui-disabled MuiMenuItem-root Mui-disabled MuiMenuItem-gutters css-nl63xq-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="exportMenuItem"
            role="menuitem"
            tabindex="-1"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1xrrpwr-MuiTypography-root"
            >
              Export
            </p>
          </li>
          <li
            aria-disabled="true"
            class="MuiButtonBase-root MuiMenuItem-root Mui-disabled MuiMenuItem-gutters Mui-disabled MuiMenuItem-root Mui-disabled MuiMenuItem-gutters css-nl63xq-MuiButtonBase-root-MuiMenuItem-root"
            data-testid="resetMenuItem"
            role="menuitem"
            tabindex="-1"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1xrrpwr-MuiTypography-root"
            >
              Reset
            </p>
          </li>
        </ul>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
